saltstack/base/salt/cloud-init/etc/scripts/extract.sh

```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env bash
+
+date >> /var/log/extract.log
+echo "OP: Extracting files ..." >> /var/log/extract.log
+
+date >> /var/log/extract.log
+echo "OP: Extracting salt..." >> /var/log/extract.log
+tar -xvzf /opt/salt_3000.2-archive.tar.gz -C /opt >> /var/log/extract.log
+date >> /var/log/extract.log
+
+date >> /var/log/extract.log
+echo "OP: Extracting td-agent..." >> /var/log/extract.log
+tar -xvzf /opt/td-agent-archive.tar.gz -C /opt >> /var/log/extract.log
+date >> /var/log/extract.log
+
+date >> /var/log/extract.log
+echo "OP: Extracting python3.6 lib64..." >> /var/log/extract.log
+tar -xvzf /usr/lib64/python3.6-archive.tar.gz -C /usr/lib64 >> /var/log/extract.log
+date >> /var/log/extract.log
+
+date >> /var/log/extract.log
+echo "OP: Extracting python3.6 lib..." >> /var/log/extract.log
+tar -xvzf /usr/lib/python3.6-archive.tar.gz -C /usr/lib >> /var/log/extract.log
+date >> /var/log/extract.log
+
+date >> /var/log/extract.log
+echo "OP: Extracting python2.7(overwrite) lib64..." >> /var/log/extract.log
+tar -xvzf /usr/lib64/python2.7-archive.tar.gz -C /usr/lib64 >> /var/log/extract.log
+date >> /var/log/extract.log
+
+date >> /var/log/extract.log
+echo "OP: Extracting python2.7(overwrite) lib..." >> /var/log/extract.log
+tar -xvzf /usr/lib/python2.7-archive.tar.gz -C /usr/lib >> /var/log/extract.log
+date >> /var/log/extract.log
+
+echo OP: Done extracting all files.
+
+
+
+
+# Notes on why this is done.
+# - The way EC2 launches instances, a lot of files end up being downloaded from S3? on-demand.
+# 
+# What shows up in scale-up timing - On the first start of these services
+# - salt-minion takes 20 seconds to launch after invoking the command
+# - ipa-install takes 20 seconds
+# - td-agent takes 10 seconds
+# - etc.
+#
+# Downloading single large archives seems to be faster than individual files. (python-* - 2000+ files each, salt - 3000+, td-agent - 50K+ iirc)
+# Not all of these files are accessed while starting up, but a fair number are.
+# Using an archive in this manner, for a single node leads to the following in terms of startup cost.
+#
+# extracting the files -> +5 seconds
+# ipa install -> -5 to -13 seconds
+# salt startup -> -15 seconds
+# td-agent startup -> -5 seconds (td-agent - we likely need to purge unnecessary rubbish (man pages, unneeded plugins) in the image baking process)
```

Comment:
this means roughly 30sec gain totally
my opinion this isn't worth it. the possibility it breaks something vs the 30sec is just too big risk to take

Reply:
30s is one of the bigger chunks that can be saved. Most other places are 1, 5, 10s - some of which get removed in the CB section of this patch. Could you please help me understand the risks here?
This change is limited to the image only - so is fairly easy to undo if required. It is configured to run via cloud-init once-per-instance, and I've verified that stop/start does not run it again. cloud-init has to complete before any of these components are used.
It does add some potential maintenance. e.g. if the salt or python version is changed, this may require some changes as well; my guess would be this isn't the only place.

Reply:
Removed this for now. Will post a separate PR.